py_modules/db_api.py

```diff
@@ -112,84 +165,104 @@ def get_unopened(self):
+    cursor = current_connection.execute(current_query,
+                                        (current_migration.date_string, current_migration.progressive_number))
+    value = cursor.fetchone()
+    if value:
+        if value[5] != current_migration.checksum:
+            exception_message = f'The migration {current_migration.file_name} has been executed with checksum ' \
+                                f'{value[5]} but the value {current_migration.checksum} was found.'
+            raise DifferentChecksumMigrationException(exception_message)
+        else:
+            raise AlreadyMigratedException(f'The migration {current_migration.file_name} has already been executed.')
+
+    current_query = "select * from migrations where date_millis >= ? and progressive_number >= ? " \
+                    "order by date_millis limit 1"
+    cursor = current_connection.execute(current_query, (current_migration.date_millis,
+                                                        current_migration.progressive_number))
+    value = cursor.fetchone()
+    if value:
+        raise OldStateMigrationException(
+            f'There current migration state ({value[1]}/{value[3]}) is more recent than '
+            f'{current_migration.date_string}/{current_migration.progressive_number}')
+
+
+def get_sorted_migrations(current_directory):
+    migrations = []
+    for file_name in current_directory:
+        try:
+            migrations.append(Migration(file_name))
+        except NotAMigrationException:
+            pass
+    migrations.sort(key=lambda x: (x.date_millis, x.progressive_number))
+    return migrations
+
+
+# EXECUTING MIGRATIONS
+MIGRATION_FILE_PATTERN = r'([0-9]{4}-([0][0-9](?=)|[1][0-2])-([0,1,2][0-9](?=)|[3][0,1]))-([0-9]+)_([a-zA-Z0-9_-]+)\.py'
+DB_LOCATION = ".db/app.db"
+MIGRATIONS_LOCATION = './.db/migrations'
+connection = sqlite3.connect(DB_LOCATION)
+connection.execute('create table if not exists migrations('
+                   'id integer primary key,'
+                   'date_string text,'
+                   'date_millis integer,'
+                   'progressive_number integer,'
+                   'file_name text,'
+                   'checksum text);')
+connection.commit()
+directory = os.listdir(MIGRATIONS_LOCATION)
+migrations_list = get_sorted_migrations(directory)
+for migration in migrations_list:
+    try:
+        validate_migration_execution(migration, connection)
+        connection.executescript(migration.sql_script)
+
+        query = 'insert into migrations values ( ( select coalesce(max(id), 0)+1 from migrations ), ?, ?, ?, ?, ?);'
+        connection.execute(query, (migration.date_string, migration.date_millis, migration.progressive_number,
+                                   migration.name, migration.checksum))
+    except AlreadyMigratedException:
+        pass
+
+connection.commit()
+connection.close()
+
+""" UPLOADING APP SETTINGS """
+with open(".db/mail_server.json", "r") as file:
```

Comment:
Nota: questo comando viene avviato ogni volta che l'app viene aperta. Lasciamo così?

Reply:
Per me è indifferente. Ho solamente aggiunto un controllo esplicito per evitare che i settings venissero inseriti più volte.

Reply:
Ok lasciamo così per ora. Al massimo apriamo un issue più avanti.

Reply:
Giusto per completezza, una query SQL in minuscolo non si può vedere🙈